shell/platform/android/io/flutter/embedding/engine/plugins/FlutterPlugin.java

```diff
@@ -0,0 +1,125 @@
+ * {@link io.flutter.plugin.common.MethodChannel} as a {@link io.flutter.plugin.common.BinaryMessenger}.
+ * <p>
+ * An Android Flutter plugin may require access to app resources or other artifacts that can only
+ * be retrieved through a {@link Context}. Developers can access the application context via
+ * {@link FlutterPluginBinding#getApplicationContext()}.
+ * <p>
+ * TODO(mattcarroll): explain ActivityAware when it's added to the new plugin API surface.
+ */
+public interface FlutterPlugin {
+
+  /**
+   * This {@code FlutterPlugin} has been associated with a {@link FlutterEngine} instance.
+   * <p>
+   * Relevant resources that this {@code FlutterPlugin} may need are provided via the {@code binding}.
+   * The {@code binding} may be cached and referenced until
+   * {@link #onDetachedFromEngine(FlutterPluginBinding)} is invoked and returns.
+   */
+  void onAttachedToEngine(@NonNull FlutterPluginBinding binding);
+
+  /**
+   * This {@code FlutterPlugin} has been removed from a {@link FlutterEngine} instance.
+   * <p>
+   * The {@code binding} passed to this method is the same instance that was passed in
+   * {@link #onAttachedToEngine(FlutterPluginBinding)}. It is provided again in this method as a
+   * convenience. The {@code binding} may be referenced during the execution of this method, but
+   * it must not be cached or referenced after this method returns.
+   * <p>
+   * {@code FlutterPlugin}s should release all resources in this method.
+   */
+  void onDetachedFromEngine(@NonNull FlutterPluginBinding binding);
+
+  /**
+   * Resources made available to all plugins registered with a given {@link FlutterEngine}.
+   * <p>
+   * The {@code FlutterPluginBinding}'s {@code flutterEngine} refers to the {@link FlutterEngine}
+   * that the associated {@code FlutterPlugin} is intended to apply to. For example, if a
+   * {@code FlutterPlugin} needs to setup a communication channel with its associated Flutter app,
+   * that can be done by wrapping a {@code MethodChannel} around
+   * {@link FlutterEngine#getDartExecutor()}.
+   * <p>
+   * A {@link FlutterEngine} may move from foreground to background, from an {@code Activity} to
+   * a {@code Service}, and {@code FlutterPluginBinding}'s {@code lifecycle} generalizes those
+   * lifecycles so that a {@code FlutterPlugin} can react to lifecycle events without being
+   * concerned about which Android Component is currently holding the {@link FlutterEngine}.
+   * TODO(mattcarroll): add info about ActivityAware and ServiceAware for plugins that care.
+   */
+  class FlutterPluginBinding {
+        private final Context applicationContext;
+        private final FlutterEngine flutterEngine;
+        private final Lifecycle lifecycle;
+
+        public FlutterPluginBinding(
+            @NonNull Context applicationContext,
+            @NonNull FlutterEngine flutterEngine,
+            @NonNull Lifecycle lifecycle
```

Comment:
This means we are forcing plugins to depend on some androidx dependency?

This might lead to painful issues in the Flutter ecosystem if this dependency ever gets a non backward compatible update. Once some of the plugins upgrade to the new version, we will end up with conflicts that are hard to deal with(and that our constraint resolver is unaware of).

I suggest we play it safe and maintain our own equivalent for lifecycle as part of the Android embedding.

Reply:
Can you describe "our own equivalent"? I don't think there can be an equivalent because Lifecycle hooks into Lifecycle-aware components. I don't think we can do that. 

I also think it's important that we embrace, where appropriate, the way that Android developers are building their apps. Lifecycle is a very popular API and it is relevant to many areas of an Android app. I think I'd need to see a very strong case for why we should avoid this particular API. 

I'm not aware of any specific compatibility issues at this point in time, and I don't know of any reason to think that this API would somehow evolve in problematic ways where other Android APIs would not. Even our Fragment definition comes from a support library, which also has an androidx analog. I think we're simply in a world now where we need to deal with the fact that androidx and support libraries both exist. Avoiding said libraries will hamstring us and make Flutter's Android embedding look antiquated as teams investigate its applicability.

In any event, I'd like to see some real customers demonstrate issues with this. I think the API surface is too important to assume that it won't work. Do you know of any customers who would likely reveal compatibility issues with this library? If not, I've got customers on a migration list anyway, and we can see if they run into any issues.

Reply:
In general we have been burnt multiple times by plugins having dependencies(having conflicts for transitive dependency) and have learned the lesson to try and avoid having dependencies to plugins as much as we can.

Using Lifecycle here will force plugins to have add this androidx depdency, the same dependency for multiple plugins. While this might never manifest into a problem if that dependency never gets a major version update, if it will get a breaking change the entire ecosystem will likely be very broken. The fact that this is a dependency that pub is unaware of even makes it worst.

I agree, there is no issue with it today, but I'm concerned that doing this will move the entire ecosystem to a very dangerous position.

Reply:
The engine repo may depend on the base Android SDK. It should not depend on the support libraries, because by definition those aren't part of Android, and we don't want to force people to take on specific dependencies beyond Flutter (which they chose) and the Android SDK (which they have to as part of writing Android apps).